### PR TITLE
[core] Sort partition before batch insert to avoid out-of-memory error.

### PR DESCRIPTION
**Pull request by leaves12138**
### Purpose

Sort partitions before batch insert without partition specified (e.g. insert overwrite and insert from a batch table source).

### API and Format

Add options `sort-partition-before-batch-insert` default true. If set this to false, will disable this feature.

No influence to user if no bugs.

### CODE REVIEW

File: paimon-core/src/main/java/org/apache/paimon/sort/ExternalRowSortBuffer.java [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+ *     a, 3, 100
+ * </pre>
+ *
+ * <p>We don't need to care about constructing the row: KEY_f2, f0, f1, f2, Then sort by KEY_f2,
+ * Then drop KEY_f2
+ *
+ * <p>The whole process is finished by this class. We just need to tell it which columns are sorted
+ * by.
+ */
+public class ExternalRowSortBuffer implements SortBuffer {
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/sort/ExternalRowSortBuffer.java:69]: Merge this class into `BatchBucketSorter`.

File: paimon-core/src/main/java/org/apache/paimon/table/sink/TableWriteImpl.java [hunk 2]
```diff
@@ -193,7 +289,27 @@ public void notifyNewFiles(
...
+            // flush external row sort buffer.
+            MutableObjectIterator<InternalRow> iterator = externalRowSortBuffer.sortedIterator();
+            BinaryRow lastPartition = new BinaryRow(partitionFields.length);
+            InternalRow sortedRow;
+            int lastBucket = -1;
+            while ((sortedRow = iterator.next()) != null) {
+                int bucket = bucket(sortedRow);
+                SinkRecord record = toSinkRecord(toOriginRow(sortedRow), bucket);
+                if (!lastPartition.equals(record.partition()) || lastBucket != bucket) {
+                    commitMessages.addAll(write.prepareCommit(waitCompaction, commitIdentifier));
```
[the remark this answers is not in the document]
Comment (leaves12138) [reply paimon-core/src/main/java/org/apache/paimon/table/sink/TableWriteImpl.java:263]: done

**Comment by leaves12138** [comment]
Thanks for review, I have modified my code according to your suggestions.

**Comment by JingsongLi** [comment]
The sorting we added may not be appropriate. We need to dynamically determine the number of writers and then dynamically determine whether to use sorting.

**Comment by leaves12138** [comment]
Finish this, we don't need to sort partition, because we can spill write buffer.

**Pull request closed without merging**